lib/remoteproc/remoteproc.c

```diff
@@ -32,7 +32,7 @@ remoteproc_check_fw_format(const void *img_data, size_t img_len)
 static struct remoteproc_mem *
 remoteproc_get_mem(struct remoteproc *rproc, const char *name,
 		   metal_phys_addr_t pa, metal_phys_addr_t da,
-		   void *va, size_t size)
+		   void *va, size_t size, struct remoteproc_mem *buf)
```

Comment:
Could you explain why you need to add remoteproc_mem parameter as the structure is already returned by the function?
look to me that you need a copy of the remoteproc_mem structure.

why not having remote proc ops like this:
```
int (*get_mem)(struct remoteproc *rproc, const char *name,
		       metal_phys_addr_t pa, metal_phys_addr_t da,
		       void *va, size_t size, struct remoteproc_mem **mem);

```
and then you could call the ops like this
```
		ret = rproc->ops->get_mem(rproc, name, pa, da, va, size, &mem);
		if (ret == 0)
			return mem;

```

Reply:
Because the porting layer normally has the different representation of memory region. If we use the second prototype, teh porting layer have to maintain another list which's type is remoteproc_mem and waste the memory.
It's also a common usage in the standard library design, for example:
https://en.cppreference.com/w/c/chrono/gmtime
```
struct tm *gmtime_r(const time_t *timep, struct tm *result);
```

Reply:
Please, could you provide link to associated porting layer and user code that i understand how you use this?

If you want to propose this kind of API i think you also need to implement this for get_mem ops to allow a porting layer relying on a list. With an appropriate description of the ops that describes the expected behavior.

Reply:
> Please, could you provide link to associated porting layer and user code that i understand how you use this?
> 

Here(note I haven't convert the implementation from mmap to get_mem yet):
https://github.com/apache/incubator-nuttx/blob/master/drivers/rptun/rptun.c#L210-L257
The mapping is done in these two function:
https://github.com/apache/incubator-nuttx/blob/master/drivers/rptun/rptun.c#L647-L668
https://github.com/apache/incubator-nuttx/blob/master/drivers/rptun/rptun.c#L670-L691
The table structure:
https://github.com/apache/incubator-nuttx/blob/master/include/nuttx/rptun/rptun.h#L261-L266

> If you want to propose this kind of API i think you also need to implement this for get_mem ops to allow a porting layer relying on a list. With an appropriate description of the ops that describes the expected behavior.

Done.

Reply:
Thanks for the links! that help to clarify my understanding of the use case!